<hr style="border: 2px solid green;" />

Hello, Avi !🤚 

My name is Kseniya Vishnevskaya. I'm going to review your project 😊 

✏️You can find my comments in <font color='green'>green</font>, <font color='orange'>yellow</font> or <font color='red'>red</font> boxes like this:

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    I'm using green color if everything is done perfectly.</div>       
<br>
<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    if I can give some recommendations</div>   
<br>
<div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    if the block requires some corrections. Work can't be accepted with the red comments</div>
 
Further in the file you can see my comments, try to take them into account when performing the next tasks. **Please, don't delete them.** :)

Let's work on the project in dialogue: if you change something in the project or respond to my comments, write about it. It will be easier for me to track the changes if you highlight your comments:

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> <b>Student comment:</b> In the blue boxes you can leave comments and ask questions.</div>


<hr style="border: 2px solid green;" />

### <font color='orange'>General feedback</font>

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    Thank you for submitting your project 🙂  I'm glad to say that you executed your project really well. I especially liked the way you analysed the missing values, wrote functions and using the quartile method. But there are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix. You've already done a lot, I am sure you'll take care of it!</div>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    There are some comments for further improvement. They don't mean that you did anything wrong, but they're things we wanted to point out that can help you further develop your skills. 

And also I leave several links that can help you to make your further projects even better:

* Text formatting:
<a>https://medium.com/analytics-vidhya/the-ultimate-markdown-guide-for-jupyter-notebook-d5e5abf728fd</a>

*  How to deal with missing data:
<a>https://towardsdatascience.com/how-to-deal-with-missing-data-in-python-1f74a9112d93</a>

*  How to edit text in jupyter notebook using Markdown:
<a>https://www.earthdatascience.org/courses/intro-to-earth-data-science/file-formats/use-text-files/format-text-with-markdown-jupyter-notebook/</a> 
    
   
</div>

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing V2:</b><br>
   I really appreciate the corrections you sent in! Thanks for taking the time to do so. However, I have a few tiny corrections to help you with. Could you take a second glance at it? One more time and you'll have it! 😊</div>

<font color='orange'>**It's strange that my comments I sent yesterday were not saved in the project. I hope you got them after all. Maybe there was some technical problem and that's why I can't see them now.**  </font>

<div style="border:solid green 2px; padding: 20px"> <b>Success V3:</b><br>
    Thank you for the update! Great job, your project has been accepted. Congratulations! 😊</div> 

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A **credit scoring** is used to evaluate the ability of a potential borrower to repay their loan.

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
 Well done for not forgetting the title and description of the project task :). But it is also important to include a description of the data to be worked with.
    </div> 

<div class="alert alert-info"> <b> What do you mean by that? where to include it? btw if by "title and description of the project task" you meant whats writen wbove your remark, it was already there..</b>

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
I praised you for having the title and description of the project task , but recommended adding a data description, like this:   
   
   **Description of the data**

    
   `children` - the number of children in the family
   
   `days_employed` - how long the customer has been working
   
   `dob_years` -  the customer’s age
   
   `education` - the customer’s education level
 
   etc.</div>   

## Open the data file and have a look at the general information. 

In [ ]:
#importing pandas
import pandas as pd

In [1]:
# reading the file
try:
    credit_score = pd.read_csv(r'C:\Users\Avi\Desktop\D.A\Practicum\Projects\1st project\credit_scoring_eng.csv')
except:
    credit_score = pd.read_csv('/datasets/credit_scoring_eng.csv')

# copying the data so it will load faster
credit_score_raw = credit_score.copy()

# checking the columns for nulls and type and other info
credit_score_raw.info()

# checking the first 10 rows of the data to see the values
credit_score_raw.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
Best practice: for convenience, it is common to keep all libraries used in code in a separate cell so that the person who will be using your project can set up the correct environment on his computer.
    
</div>   

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> You mean that i should do the import pandas in a separate cell?

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
    Yes, it's better to use separate cell for importing all libraries, including pandas :)</div>   

In [2]:
# using describe to get more general info about the values
credit_score_raw.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

<s><div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>

   The `describe()` function is very useful in getting to know the data. In this case, it looks much more indicative if the output is expanded by `df.describe().T`. In future projects, the `include` argument may come in handy - `df.describe(include = ['object']` is used to analyse categorical values.
    
</div>  

In [3]:
credit_score_raw.describe().T

count          mean            std           min  \
children          21525.0      0.538908       1.381587     -1.000000   
days_employed     19351.0  63046.497661  140827.311974 -18388.949901   
dob_years         21525.0     43.293380      12.574584      0.000000   
education_id      21525.0      0.817236       0.548138      0.000000   
family_status_id  21525.0      0.972544       1.420324      0.000000   
debt              21525.0      0.080883       0.272661      0.000000   
total_income      19351.0  26787.568355   16475.450632   3306.762000   

                           25%           50%           75%            max  
children              0.000000      0.000000      1.000000      20.000000  
days_employed     -2747.423625  -1203.369529   -291.095954  401755.400475  
dob_years            33.000000     42.000000     53.000000      75.000000  
education_id          1.000000      1.000000      1.000000       4.000000  
family_status_id      0.000000      0.000000      1.000000       4.000000  
debt                  0.000000      0.000000      0.000000       1.000000  
total_income      16488.504500  23202.870000  32549.611000  362496.645000

In [4]:
credit_score_raw.describe(include = ['object'])

education family_status gender income_type           purpose
count                 21525         21525  21525       21525             21525
unique                   15             5      3           8                38
top     secondary education       married      F    employee  wedding ceremony
freq                  13750         12380  14236       11119               797

### Conclusion

Most of the columns seem to have no nulls and all are of the right type.

There are missing values in total_income and days_employed columns and the number of missing values are the same.

The days_employed column also has some illogical values like negative days and too long periods of employment.
Since days_employed has too long periods of employment that no one in this century is able to do, and it isnt a relevant value ill probably disregard this column and wont fill the NaNs in it.

The education column needs to be lower cased, ill do it to all object columns (maybe except gender column) just in case.

There is probably a mistake in the children column like having negative number of children and should check the customer that has 20 children (maybe s/he re-married a couple times and/or had children out of marriage? (; ) lets investigate.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    You've done a great job of sorting out the data description and writing detailed conclusions. Well done! 😊</div> 

## Data preprocessing

### Processing missing values

In [5]:
# Identifying missing values
credit_score_raw.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

It can be clearly seen that 2 columns have the exact same number of nulls.

The missing value types that were detected are floats.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    It's great that you show which columns have missing values in our dataset :) </div>       


Now ill organize the table a bit:

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> 
    lets see what percentage of negatives we have in days_employed column:

In [6]:
# finding the number of negative values in days_employed column
negative_values = credit_score_raw['days_employed'][credit_score_raw['days_employed'] < 0].count()
# calculating the percentage of them out of all the rows
percentage_of_negatives = negative_values / len(credit_score_raw) * 100
# orinting the rounded percentage
percentage_of_negatives.round(3)

73.895

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> 
Seems the majority of the values are negatives, since it is not possible to work negative days, i assume that the negatives are a typo, maybe the one made this data in a file intended it to be a dash and not a minus for example.

In [7]:
# saving to organized table
orgnzd_credit_score = credit_score_raw

In [8]:
# lower casing object columns
orgnzd_credit_score['education'] = orgnzd_credit_score['education'].str.lower()
orgnzd_credit_score['family_status'] = orgnzd_credit_score['family_status'].str.lower()
orgnzd_credit_score['income_type'] = orgnzd_credit_score['income_type'].str.lower()
orgnzd_credit_score['purpose'] = orgnzd_credit_score['purpose'].str.lower()

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Very well that you converted the columns to lower case using <code>str.lower()</code></div>  

Since its not possible to work negative days and have negative children lets fix the negatives typos, j.i.c i'll apply to all relevant columns

In [9]:
# removing negatives from the relevant columns
orgnzd_credit_score['children'] = orgnzd_credit_score['children'].abs()
orgnzd_credit_score['days_employed'] = orgnzd_credit_score['days_employed'].abs()
orgnzd_credit_score['dob_years'] = orgnzd_credit_score['dob_years'].abs()
orgnzd_credit_score['total_income'] = orgnzd_credit_score['total_income'].abs()
orgnzd_credit_score.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.543275   66914.728907     43.293380      0.817236   
std        1.379876  139030.880527     12.574584      0.548138   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000     927.009265     33.000000      1.000000   
50%        0.000000    2194.220567     42.000000      1.000000   
75%        1.000000    5537.882441     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
   
   I agree that these values are wrong and should be processing but in that case it is worth checking what percentage of this data is in the dataframe (especially `days_employed`). Could you please check it and assume why there might be errors in the data?</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px"> 
    Done above (for some reason when i try to do it after i removed the negatives in the `new` table it doesnt find the negatives in the `raw data`)

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
    Good job!</div>       

Lets find the rows with 20 children, 401755.400475 days_employed and 362496.645000 total_income:

In [10]:
# trying to find the row of the max children:
orgnzd_credit_score.query('children == 20')

children  days_employed  dob_years            education  education_id  \
606          20     880.221113         21  secondary education             1   
720          20     855.595512         44  secondary education             1   
1074         20    3310.411598         56  secondary education             1   
2510         20    2714.161249         59    bachelor's degree             0   
2941         20    2161.591519          0  secondary education             1   
...         ...            ...        ...                  ...           ...   
21008        20    1240.257910         40  secondary education             1   
21325        20     601.174883         37  secondary education             1   
21390        20            NaN         53  secondary education             1   
21404        20     494.788448         52  secondary education             1   
21491        20     173.954460         27  secondary education             1   

         family_status  family_status_id gender income_type  debt  \
606            married                 0      M    business     0   
720            married                 0      F    business     0   
1074           married                 0      F    employee     1   
2510   widow / widower                 2      F    employee     0   
2941           married                 0      F    employee     0   
...                ...               ...    ...         ...   ...   
21008          married                 0      F    employee     1   
21325          married                 0      F    business     0   
21390          married                 0      M    business     0   
21404          married                 0      M    business     0   
21491          married                 0      F    employee     0   

       total_income                                   purpose  
606       23253.578                     purchase of the house  
720       18079.798                           buy real estate  
1074      36722.966                      getting an education  
2510      42315.974  transactions with commercial real estate  
2941      31958.391                              to buy a car  
...             ...                                       ...  
21008     21363.842                              to own a car  
21325     16477.771                         profile education  
21390           NaN               buy residential real estate  
21404     25060.749          transactions with my real estate  
21491     24883.344                  buying a second-hand car  

[76 rows x 12 columns]

Seems there are 76 rows with this number of children which is not probable at all so its probably a typo and supposed to be 2.
But lets check if its the only irregular value or maybe  for example there are people with 15 kids as well:

In [11]:
orgnzd_credit_score['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

Yes, most probably a typo. So lets change it to 2.

In [12]:
orgnzd_credit_score.loc[orgnzd_credit_score['children'] == 20, 'children'] = 2
orgnzd_credit_score['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Well done! </div>       

ok now lets do the same check for the other 2 columns based on describe:

In [13]:
# trying to find the row of the max days_employed:
# orgnzd_credit_score[orgnzd_credit_score['days_employed'] == 401755.400475]
# orgnzd_credit_score[orgnzd_credit_score.days_employed == 401755.400475]
#orgnzd_credit_score.query('days_employed == 401755.400475')
orgnzd_credit_score[orgnzd_credit_score['days_employed'] > 401755]

children  days_employed  dob_years            education  education_id  \
6954         0  401755.400475         56  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
6954  widow / widower                 2      F     retiree     0   

      total_income             purpose  
6954     28204.551  housing renovation

<s><div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
   
   Here you can find the boundaries of the values that can be considered abnormal and use a slice, like this: `orgnzd_credit_score[orgnzd_credit_score['days_employed'] > 401755]` </div>   

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    Thanks, but now im a bit confused why it worked.. i get the integer trick but at this point there are still NaNs in the table..

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
   I guess not all numbers are reflected after the comma, so eventually the equality was not fulfilled and the desired line is not found.</div>       

Based on what i gather from the web its not possible to .loc[] or query or any other way when there are floats and NaNs, so ill try it later.

Great, now back to the missing values:

In [14]:
# finding the rows where the values are missing because they are probably the same rows to see what they have in common
orgnzd_credit_score.loc[orgnzd_credit_score['total_income'].isna()].head(50)

children  days_employed  dob_years            education  education_id  \
12          0            NaN         65  secondary education             1   
26          0            NaN         41  secondary education             1   
29          0            NaN         63  secondary education             1   
41          0            NaN         50  secondary education             1   
55          0            NaN         54  secondary education             1   
65          0            NaN         21  secondary education             1   
67          0            NaN         52    bachelor's degree             0   
72          1            NaN         32    bachelor's degree             0   
82          2            NaN         50    bachelor's degree             0   
83          0            NaN         52  secondary education             1   
90          2            NaN         35    bachelor's degree             0   
94          1            NaN         34    bachelor's degree             0   
96          0            NaN         44  secondary education             1   
97          0            NaN         47    bachelor's degree             0   
120         0            NaN         46  secondary education             1   
121         0            NaN         29    bachelor's degree             0   
135         0            NaN         27  secondary education             1   
141         0            NaN         39  secondary education             1   
145         0            NaN         62  secondary education             1   
174         0            NaN         55    bachelor's degree             0   
181         0            NaN         26  secondary education             1   
189         1            NaN         30  secondary education             1   
205         1            NaN         31    bachelor's degree             0   
220         1            NaN         23         some college             2   
241         0            NaN         47  secondary education             1   
242         0            NaN         58  secondary education             1   
247         1            NaN         60    bachelor's degree             0   
250         0            NaN         54    bachelor's degree             0   
264         2            NaN         40  secondary education             1   
278         1            NaN         23  secondary education             1   
279         0            NaN         44  secondary education             1   
280         0            NaN         66  secondary education             1   
312         1            NaN         33  secondary education             1   
317         0            NaN         21    bachelor's degree             0   
320         0            NaN         63  secondary education             1   
321         1            NaN         34    bachelor's degree             0   
322         0            NaN         56  secondary education             1   
328         0            NaN         69  secondary education             1   
360         0            NaN         27  secondary education             1   
361         0            NaN         59  secondary education             1   
364         0            NaN         57  secondary education             1   
376         0            NaN         23         some college             2   
383         0            NaN         55    bachelor's degree             0   
389         1            NaN         31  secondary education             1   
415         0            NaN         57    bachelor's degree             0   
416         0            NaN         58  secondary education             1   
421         0            NaN         45  secondary education             1   
444         0            NaN         53  secondary education             1   
458         0            NaN         48  secondary education             1   
466         0            NaN         56  secondary education             1   

         family_status  fa

It seems that most of the rows have "secondary education" in the "education" column and 0 in debt column so maybe the clue is there.. lets check

In [15]:
# making a new table that contains all the NaN rows
credit_score_NaN_rows = orgnzd_credit_score.loc[orgnzd_credit_score['total_income'].isna()]
# making a new table that contains all the no NaN rows
credit_score_no_NaN = orgnzd_credit_score.loc[orgnzd_credit_score['total_income'].notna()]

In [16]:
# grouping the NaN table by the education column to see what kinds are in the table and how much of each there are
credit_score_NaN_rows.groupby('education')['education'].count()

education
bachelor's degree       544
primary education        21
secondary education    1540
some college             69
Name: education, dtype: int64

In [17]:
# grouping the NaN table by the debt column to see how much 0s are in there
credit_score_NaN_rows.groupby('debt')['education'].count()

debt
0    2004
1     170
Name: education, dtype: int64

In [18]:
# checking how much categories there are in 'education' column in general table
len(orgnzd_credit_score.groupby('education'))

5

Seems 4 out of 5 types of education is part of the NaN table so it (secondary education) is not the constant.

Because i cant find a pattern behind the missing values, in my opinion these missing values are MCAR so the Possible reason is some technical error (i thought that maybe it was just left empty and meant by that to be 0 but that is also out of the question since a lot of the NaN rows have employee income types and 92% of them never have been in debt so that doesnt add up).

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    I like the way you have carefully analysed the omissions in the data.</div> 

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
   You can also use the missingno library to view missing values. I'll leave a link, perhaps you'll be interested:
<a>https://github.com/ResidentMario/missingno</a>
</div>    

### Filling the missing values

I will fill the NaNs in total_income based on groups in income_type.

In [19]:
# grouping the no NaN table by income_type to see how much data i have on each group to make a median/mean of
#credit_score_no_NaN.groupby('income_type')['total_income'].count()
credit_score_no_NaN.groupby('income_type').agg({'total_income': ['count']})

total_income
                                   count
income_type                             
business                            4577
civil servant                       1312
employee                           10014
entrepreneur                           1
paternity / maternity leave            1
retiree                             3443
student                                1
unemployed                             2

if there is so little count for student and other income groups in the no NaN table, maybe they (students and others) are the majority in the NaN table, lets check:

In [20]:
# checking the count of each group of income_type ib the NaN table
#credit_score_NaN_rows.groupby('income_type')['income_type'].count()
# credit_score_NaN_rows['income_type'].value_counts()
credit_score_NaN_rows.groupby('income_type').agg({'income_type': ['count']})

income_type
                    count
income_type              
business              508
civil servant         147
employee             1105
entrepreneur            1
retiree               413

(This proved not to be the case, there is no student, unemployed, paternity / maternity leave, and only 1 entrepreneur in the NaN table, so the theory about them being the reason for the NaNs is out of the window)

Since there is not enough data about [student, unemployed, paternity / maternity leave, entrepreneur and civil servant] I think i will fill the data only for [business, employee, retiree] since they are the more significant part of the data.

Based on the info from the describe method, there are outliers in total_income column so ill fill with median and not mean.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    It is correct that you chose the median to fill in the missing values 👍</div>       

Since if there is duplicate rows it will skew the calculations of the median, i will first check for them in the table ill use.

In [21]:
# checking for duplicates in the table by which i calculate the medians to fill the NaNs.
credit_score_no_NaN.duplicated().sum()

0

thankfully seems to be no row duplicates in the table i will calculate the median by, lets check if there is a need for Stemming or Lemmatization for the income_type column (and others while we're at it) in the general table:

In [22]:
# orgnzd_credit_score['income_type'].value_counts()#.sum()
orgnzd_credit_score.groupby('income_type').agg({'income_type': ['count']})

income_type
                                  count
income_type                            
business                           5085
civil servant                      1459
employee                          11119
entrepreneur                          2
paternity / maternity leave           1
retiree                            3856
student                               1
unemployed                            2

I have only 2 entrepreneur rows in general and 1 of them is NaN. So ill think what to do with it..

In [23]:
# orgnzd_credit_score['education'].value_counts()#.sum()
orgnzd_credit_score.groupby('education').agg({'education': ['count']})

education
                        count
education                    
bachelor's degree        5260
graduate degree             6
primary education         282
secondary education     15233
some college              744

In [24]:
# orgnzd_credit_score['family_status'].value_counts()#.sum()
orgnzd_credit_score.groupby('family_status').agg({'family_status': ['count']})

family_status
                          count
family_status                  
civil partnership          4177
divorced                   1195
married                   12380
unmarried                  2813
widow / widower             960

In [25]:
# orgnzd_credit_score['gender'].value_counts()#.sum()
orgnzd_credit_score.groupby('gender').agg({'gender': ['count']})

gender
        count
gender       
F       14236
M        7288
XNA         1

XNA might be because of a third gender or incorrect information while data entry. Since there is only 1 and i wont do any gender based statistics, lets change XNA to M.

In [26]:
orgnzd_credit_score.loc[orgnzd_credit_score['gender'] == 'XNA', 'gender'] = 'M'

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
  Well done for fixing 'XNA' gender in your project.</div>`

In [27]:
orgnzd_credit_score.groupby('gender').agg({'gender': ['count']})

gender
        count
gender       
F       14236
M        7289

In [28]:
# orgnzd_credit_score['purpose'].value_counts()#.sum()
orgnzd_credit_score.groupby('purpose').agg({'purpose': ['count']})

purpose
                                           count
purpose                                         
building a property                          620
building a real estate                       626
buy commercial real estate                   664
buy real estate                              624
buy residential real estate                  607
buying a second-hand car                     479
buying my own car                            505
buying property for renting out              653
car                                          495
car purchase                                 462
cars                                         478
construction of own property                 635
education                                    447
getting an education                         443
getting higher education                     426
going to university                          496
having a wedding                             777
housing                                      647
housing renovation                           612
housing transactions                         653
profile education                            436
property                                     634
purchase of a car                            455
purchase of my own house                     620
purchase of the house                        647
purchase of the house for my family          641
real estate transactions                     676
second-hand car purchase                     489
supplementary education                      462
to become educated                           412
to buy a car                                 472
to get a supplementary education             447
to have a wedding                            774
to own a car                                 480
transactions with commercial real estate     651
transactions with my real estate             630
university education                         453
wedding ceremony                             797

Seems that there are duplicates in the purpose column, but since it doesnt affect any median calculasion and filling missing values, ill first finish filling the missing values and then continue handling the duplicates.

In [29]:
# finding total_income median for each (relevant to fill) income_type.
total_income_median = credit_score_no_NaN.groupby('income_type')['total_income'].median()
# making a variable for each.
business_income_median = total_income_median['business']
employee_income_median = total_income_median['employee']
retiree_income_median = total_income_median['retiree']
# I was very hesitant about adding the 'civil servant' group because there is only 1312 of them 
# but since it is the rest of the nulls in the column i decided to add it too.
civil_servant_income_median = total_income_median['civil servant']
# this is to make the pretty table (the count is to see clearly which income type are not relevant for medians) :P
total_income_medians = credit_score_no_NaN.groupby('income_type').agg({'total_income': ['median', 'count']})
total_income_medians

total_income       
                                  median  count
income_type                                    
business                      27577.2720   4577
civil servant                 24071.6695   1312
employee                      22815.1035  10014
entrepreneur                  79866.1030      1
paternity / maternity leave    8612.6610      1
retiree                       18962.3180   3443
student                       15712.2600      1
unemployed                    21014.3605      2

In [30]:
# finding days_employed median for each (relevant to fill) income_type.
days_employed_median = credit_score_no_NaN.groupby('income_type')['days_employed'].median()
# making a variable for each.
business_employed_median = days_employed_median['business']
employee_employed_median = days_employed_median['employee']
retiree_employed_median = days_employed_median['retiree']
# I was very hesitant about adding the 'civil servant' group because there is only 1312 of them 
# but since it is the rest of the nulls in the column i decided to add it too.
civil_servant_employed_median = days_employed_median['civil servant']
# this is to make the pretty table (the count is to see clearly which income type are not relevant for medians) :P
days_employed_medians = credit_score_no_NaN.groupby('income_type').agg({'days_employed': ['median', 'count']})
days_employed_medians

days_employed       
                                    median  count
income_type                                      
business                       1547.382223   4577
civil servant                  2689.368353   1312
employee                       1574.202821  10014
entrepreneur                    520.848083      1
paternity / maternity leave    3296.759962      1
retiree                      365213.306266   3443
student                         578.751554      1
unemployed                   366413.652744      2

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Well done for being so thorough about handling missing values. You did a great job!</div>       

#### filling the missing values in the general table by income_type

In [31]:
# checking how much nulls i have in total_income before filling them
orgnzd_credit_score['total_income'].isna().sum()

2174

In [32]:
# checking how much nulls i have in days_employed before filling them
orgnzd_credit_score['days_employed'].isna().sum()

2174

In [33]:
# Still pretty hesitant about filling civil servant by their own median, but adding them
income_types = ['business', 'civil servant', 'employee', 'retiree']
median_income_list = [business_income_median, civil_servant_income_median, employee_income_median, retiree_income_median]
median_employed_list = [business_employed_median, civil_servant_employed_median, employee_employed_median, retiree_employed_median]

def filling_nans(data, income_list, column1, column2, median_list):
    count = 0
    for i in income_list:
        mask = (data[column1] == i)
        data.loc[mask, column2] = data.loc[mask, column2].fillna(median_list[count])
        #the one below was to use on the NaN table until i found out about the current one
        #data.at[data[column1] == i, column2] = median_list[count]
        count += 1

filling_nans(orgnzd_credit_score, income_types, 'income_type', 'total_income', median_income_list)
filling_nans(orgnzd_credit_score, income_types, 'income_type', 'days_employed', median_employed_list)

# checking how much nulls i have in total_income after filling them
orgnzd_credit_score['total_income'].isna().sum()

1

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Thank you for your bright idea and out-of-the-box approach!

</div> 

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    To simplify the work you can replace the missing values with, for example, median income depending on <code>income_type</code> and <code>education_id</code> like this:
   <code> data['total_income'] = data['total_income'].fillna(data.groupby(['education_id', 'income_type'])['total_income'].transform('median')) </code></div>   

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    You mean that i can do this instead of the function right?
    and also why group by education, and how does it work when grouping by 2 parameters?

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
   
   Yes, you can use this method in your projects. It helps you group the data by two parameters - `'education_id'` and `'income_type'` (assuming that income depends on education and income type) and find the median income from the grouped data. This helps increase the accuracy of the calculations.</div>   

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
   ok, got it, hope its ok, that since i am already behind everybody on the curriculum and its only a remark ill leave it as is.

In [34]:
# checking how much nulls i have in days_employed after filling them
orgnzd_credit_score['days_employed'].isna().sum()

1

The (148 ramaining NaNs consisted of 147 civil servant and now only the) 1 NaN row remaining is entrepreneur.

(A comment relevant to when i used the hashed structure in the function:
The code works even though there is this red notice, i read a lot about what is in the link but there is nothing relevant i found there, and trying its suggestion to "Try using .loc[row_indexer,col_indexer] = value instead" doesnt work and is not relevant).

Since the entrepreneur NaN will be a hindrance, I can either drop the row, make/rename it to business, or fill it with the other entrepreneur row.

Since it has only 2 rows in total i wont fill the NaN in it with the value from the other entrepreneur row since it will skew the total_income categories in the next steps, only based on 1 value, so i wont.

And if ill make/rename it to business then later when ill look for duplicates i might get more duplicates (because the income_type will be the same, and it may be the only difference) even though its obviously a different customer.
And since i wont be using the entrepreneur group by itself anyway (because it has only 2 rows in total) i think ill fill the entrepreneur NaN with business medians.

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
   
   If you are sure everything works correctly, you can turn off warnings messages. This article can help:
        https://docs.python.org/3/library/warnings.html

But I don't see this message. Maybe you see this message when you do a project locally on your computer.</div>   

In [35]:
## changing entrepreneur rows to business rows
# orgnzd_credit_score.loc[orgnzd_credit_score['income_type'] == 'entrepreneur', 'income_type'] = 'business'

# filling the entrepreneur NaN with business medians
mask = (orgnzd_credit_score['income_type'] == 'entrepreneur')
orgnzd_credit_score.loc[mask, 'total_income'] = orgnzd_credit_score.loc[mask, 'total_income'].fillna(business_income_median)
orgnzd_credit_score.loc[mask, 'days_employed'] = orgnzd_credit_score.loc[mask, 'days_employed'].fillna(business_employed_median)

In [36]:
# checking how much nulls i have in total_income after filling entrepreneur
orgnzd_credit_score['total_income'].isna().sum()

0

In [37]:
# checking how much nulls i have in days_employed after filling entrepreneur
orgnzd_credit_score['days_employed'].isna().sum()

0

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Well done!</div>       

### Conclusion

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    Can you please remove the item numbering when you do a project locally? Note that the Yandex.Practicum environment provides automatic numbering of project items when you use # before the item name. Otherwise, the structure of the project changes a lot now. And please write the conclusion here then :)</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    ok, it just helps me orienting while looking for something.. :/

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    I identified the missing values, decided which way to fill it (by medians of total income grouped by income types), and filled them.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Great, now the project has a clear structure :) You can just use the numbering of the sections on the local computer, and delete the numbering when uploading to the Yandex.Praсtiсum environment. And thanks for the conclusions! Now everything is perfect 😊</div>       

### Data type replacement

Now lets convert the floats to integers.

In [38]:
# converting the floats to integers.
orgnzd_credit_score['total_income'] = orgnzd_credit_score['total_income'].astype('int64')
orgnzd_credit_score['days_employed'] = orgnzd_credit_score['days_employed'].astype('int64')
orgnzd_credit_score.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


<div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    Please visit this link, there is information on how to optimise data types and downsize them: 
    <a>https://vincentteyssier.medium.com/optimizing-the-size-of-a-pandas-dataframe-for-low-memory-environment-5f07db3d72e</a>
</div> 

Now that there are no NaNs and floats i can check the other 2 columns (days_employed and total_income) for their max rows based on describe:

In [39]:
orgnzd_credit_score.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.479721   67299.057003     43.293380      0.817236   
std        0.755528  139401.780440     12.574584      0.548138   
min        0.000000      24.000000      0.000000      0.000000   
25%        0.000000    1025.000000     33.000000      1.000000   
50%        0.000000    1993.000000     42.000000      1.000000   
75%        1.000000    5347.000000     53.000000      1.000000   
max        5.000000  401755.000000     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   21525.000000  
mean           0.972544      0.080883   26433.153403  
std            1.420324      0.272661   15682.756144  
min            0.000000      0.000000    3306.000000  
25%            0.000000      0.000000   17247.000000  
50%            0.000000      0.000000   22815.000000  
75%            1.000000      0.000000   31286.000000  
max            4.000000      1.000000  362496.000000

In [40]:
# trying to find the row of the max days_employed:
orgnzd_credit_score.query('days_employed == 401755')

children  days_employed  dob_years            education  education_id  \
6954         0         401755         56  secondary education             1   

        family_status  family_status_id gender income_type  debt  \
6954  widow / widower                 2      F     retiree     0   

      total_income             purpose  
6954         28204  housing renovation

In [41]:
# trying to find the row of the max total_income:
orgnzd_credit_score.query('total_income == 362496')

children  days_employed  dob_years          education  education_id  \
12412         0           1477         44  bachelor's degree             0   

      family_status  family_status_id gender income_type  debt  total_income  \
12412       married                 0      M    business     0        362496   

                  purpose  
12412  housing renovation

Seems ok, the max value appears only once.

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    
   Please note that there are many more abnormalous values. Here you output only the maximum values of `days_employed`. But if you recalculate days in years, there are such values in the data that a person worked more than a person can live years.</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    but i addressed it later, here is only converting the types, isnt it so?

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
    
   I see you've assumed it's not days of work, but hours. This block refers to preprocessing data, it is better to denote it here. There are `days_employed` values of about 300 000 hours in the data. It means that some people have been working about 37 500 days for 8 hours, or over 100 years... 🤔 </div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    
   hmm you are right! i havent thought of it.. forgot when doing the calculation that humans usually dont work 24/7 ahah.. but then i dont know what can it be.. it was logical to count employment time by hours, but less so by minutes, its too precise.. but still plausible i guess.. the minimal value of the column is 24, i guess its possible some worked for 24 minutes until deciding its not for them and left, but the data is about the whole time a customer was employed and unless that customer is like 18 he should have worked more time in his life..
    
   So ill check now all the rows with `days_employed` < 60,000 (assuming its minutes employed and that they worked for about 5 hours a day [because they are young] for 200 days) and i expect to see that they are students or unemployed and that they are not more than 20 or so years in dob_years.

In [42]:
# finding the rows where days_employed is less than 60000 and looking at the info in children, dob_years and total_income with describe
orgnzd_credit_score.query('days_employed < 60000').describe()

children  days_employed     dob_years  education_id  \
count  17667.000000   17667.000000  17667.000000  17667.000000   
mean       0.563876    2283.415973     39.852097      0.796174   
std        0.794548    2198.421547     10.670217      0.553849   
min        0.000000      24.000000      0.000000      0.000000   
25%        0.000000     843.500000     32.000000      0.000000   
50%        0.000000    1574.000000     39.000000      1.000000   
75%        1.000000    2939.500000     48.000000      1.000000   
max        5.000000   18388.000000     75.000000      4.000000   

       family_status_id          debt   total_income  
count      17667.000000  17667.000000   17667.000000  
mean           0.969718      0.086263   27484.080263  
std            1.442329      0.280759   16161.021305  
min            0.000000      0.000000    3418.000000  
25%            0.000000      0.000000   18004.000000  
50%            0.000000      0.000000   23717.000000  
75%            1.000000      0.000000   32285.500000  
max            4.000000      1.000000  362496.000000

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    
   so here it can be seen that in a table of customers that supposedly only worked up to 200 (assuming 5 hour) working days (the actual maximal output is 18388 "minutes" which equals to even less 61 days) in their life(!), there are at least some of them with 5 children, 75 dob_years, and 362,496 in total_income (which is the max for the whole table!). Since it ilogical that a customer that is 75 years in the bank worked up to 60 days in his life (and it can also be seen in the rest of the table like that the median of dob_years is 39 which is also illogical for up to 60 days of work) i discard the possibility that it is minutes as well. I would drop the column, but since it doesnt pose any issue i will leave it and tell the ones responsible for the data that it doesnt make sense and ask them if they know what happened.

<div style="border:solid orange 2px; padding: 20px"> <b>Remarks V3:</b><br>
    
  You did a great job and did an in-depth analysis of the "days_employed" column. It's great that you looked at this data from different angles. How could this problem still be solved:

    
   Roughly calculate how many maximum years the person could work to his current age (if he is 60 years old, then 60-18=42 years). The working days are approximately 257. 257*42=10794 days at most could work in his lifetime. Think of a conditional error, let it be 1000 (if let's say he is a workaholic and worked all the time or started working at an earlier age. 10794 + 1000 = 11794. Compare this number with the value in the `days_employed` column. If 11794 < days_employed, fill in NaN. Then fill in the missing values with the median `days_employed`, grouping by `dob_years`.

</div>

### Conclusion

Now whats left to do is to `([no need to merge now] merge credit_score_NaN_rows table with credit_score_no_NaN table, then i'll)` find the duplicate rows with DataFrame.duplicated(subset=all columns except total_income and days_employed, keep='first')
[except those columns because i filled them myself based on the income_type so it doesnt matter]

And then the Stemming will begin.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    You converted the data to the right types :) Thanks for detailed conclusion. </div> 

### Processing duplicates

In [43]:
## searching for duplicated rows in the general table by all columns except both total_income and days_employed
ten_columns = ['children', 'dob_years', 'education', 'education_id', 'family_status', 'family_status_id', 'gender', 'income_type', 'debt', 'purpose']
orgnzd_credit_score.duplicated(subset=ten_columns).sum()

3654

In [44]:
# searching for duplicated rows in the general table by all columns.
orgnzd_credit_score.duplicated().sum()

71

So that means there are rows (aka people) with exactly the same parameters, with just total_income and/or days_employed as a differentiator.
Lets try to check if the ramaining 71 dups' are coming from the NaN rows where there is also no total_income and days_employed values:

In [45]:
# searching for duplicated rows in the NaN table by all columns (except both total_income and days_employed because they are NaNs)
credit_score_NaN_rows.duplicated().sum()

71

indeed they are, so all the 71 dups were the rows that were NaNs before.

So based on this most of the duplicates are surfacing when we dont compare by all columns.
To check that its not just because in the NaN table there is more chance for a mistake, lets check the no NaN table:

In [46]:
# searching for duplicated rows in the no NaN table by all columns except both total_income and days_employed
credit_score_no_NaN.duplicated(subset=ten_columns).sum()

3019

In [47]:
# searching for duplicated rows in the no NaN table by all columns.   ## except either total_income/days_employed
credit_score_no_NaN.duplicated().sum()

0

Based on the checks i did:

***`if`*** searching for duplicated rows in the **`no NaN`** table by all columns except total_income and days_employed returns **`3019`** duplicates, while searching for duplicated rows in the **`no NaN`** table by all columns (except **`either`** total_income/days_employed) returns **`0`** duplicates,

***`then`*** it means there are rows that are identical in every column except the total_income or days_employed (and that, even before stemming and lemmatizing the purpose column)! And that in turn means that we cannot remove the rows which do have NaNs based only on all the other rows... because it proves that we have customers that are the same in each category but earn different amount and/or employed a different amount of time, but all else is the same.

(3654 rows consist of 3019 rows in the no NaN table + 71 rows in the NaN table + 564 rows that are in one table only once and another time in the other table, and when both tables together, like in the general table, they [the duplicates] emerge).

But after consulting with Anastasia and because its not a real data but only a simulation ill drop these "duplicates".

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Thanks for the in-depth analysis of the duplicates. Very good!</div> 

In [48]:
# checking the table before droping duplicates
#orgnzd_credit_score.shape[0]
orgnzd_credit_score.duplicated().sum()

71

In [49]:
# removing duplicate rows and reseting the index
orgnzd_credit_score = orgnzd_credit_score.drop_duplicates().reset_index(drop=True)
#orgnzd_credit_score.reset_index(drop=True, inplace=True)
#orgnzd_credit_score.shape[0]

In [50]:
# checking the table after droping duplicates
orgnzd_credit_score.duplicated().sum()

0

In [51]:
# checking that the index reset
orgnzd_credit_score

children  days_employed  dob_years            education  education_id  \
0             1           8437         42    bachelor's degree             0   
1             1           4024         36  secondary education             1   
2             0           5623         33  secondary education             1   
3             3           4124         32  secondary education             1   
4             0         340266         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21449         1           4529         43  secondary education             1   
21450         0         343937         67  secondary education             1   
21451         1           2113         38  secondary education             1   
21452         3           3112         38  secondary education             1   
21453         2           1984         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21449  civil partnership                 1      F    business     0   
21450            married                 0      F     retiree     0   
21451  civil partnership                 1      M    employee     1   
21452            married                 0      M    employee     1   
21453            married                 0      F    employee     0   

       total_income                  purpose  
0             40620    purchase of the house  
1             17932             car purchase  
2             23341    purchase of the house  
3             42820  supplementary education  
4             25378        to have a wedding  
...             ...                      ...  
21449         35966     housing transactions  
21450         24959        purchase of a car  
21451         14347                 property  
21452         39054        buying my own car  
21453         13127             to buy a car  

[21454 rows x 12 columns]

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    Please don't forget to write conclusion in this module.</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    I identified the duplicates, removed them, and reset the index.

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
    👍 </div>       

#### Stemming

Stemming the purpose column

Since there are to much inique values in the purpose column while they really are the same pruposes that are phrased differently, I will stemm it to be only one word that represents each loan purpose: house, wedding, car and education.

In [52]:
# importing stemming tool
from nltk.stem import SnowballStemmer
english_stemmer = SnowballStemmer('english')

In [53]:
# writing a function to stemm the purpose column
def stemming_purpose(word_list, categorie):
    for i in word_list:
        for query in orgnzd_credit_score['purpose']:
            for word in query.split(' '):
                stemmed_word = english_stemmer.stem(word)
                if stemmed_word == english_stemmer.stem(i):
                    orgnzd_credit_score.loc[orgnzd_credit_score['purpose'] == query, 'purpose'] = categorie
                else:
                    continue

In [54]:
# ordering all car related loans
car_list = ['car']
stemming_purpose(car_list, 'car')

# ordering all wedding related loans
wedding_list = ['wedding']
stemming_purpose(wedding_list, 'wedding')

# ordering all house related loans
house_list = ['house', 'property', 'estate']
stemming_purpose(house_list, 'house')

# ordering all education related loans
education_list = ['education', 'university']
stemming_purpose(education_list, 'education')

In [55]:
# checking to see if it worked well
orgnzd_credit_score['purpose'].value_counts()

house        10811
car           4306
education     4013
wedding       2324
Name: purpose, dtype: int64

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   The stemming is correct and you have written the function correctly to categorise the purpose of the loan. </div>   


<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
   hushing the removal of the stemmed duplicates

In [56]:
# checking for duplicates again after stemming
#orgnzd_credit_score.duplicated().sum()

In a real data set it wouldnt be possible (in case like this of a bank where there are relatively not so much people) but here i knew there would be more duplicates (since there are rows that are identical in all columns except the NaN columns to begin with and i filled the NaNs with medians) after stemming because i remove the differentiative phrases that the one who made this dataset randomly assigned to rows. but in a real dataset about customers this differentiation in phrasing supposed to point that its a different customer, and in a real dataset there would not be 2 exactly the same customers (with exactly the same loan purpose, time in the bank, income, num of children etc) except for the way they phrased the purpose . Hope i am understandable :)

But again since its the given data ill remove those as well.  

In [57]:
# removing duplicate rows and resetting the index
#orgnzd_credit_score = orgnzd_credit_score.drop_duplicates().reset_index(drop=True)

In [58]:
# checking for duplicates again after removing
#orgnzd_credit_score.duplicated().sum()

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
    I understand your idea but here it is better not to delete the data, because formally we processed them in the previous block. You could accidentally delete data that are not really duplicates, but became duplicates after filling with median values and after changing the loan purpose. This is not entirely correct. Could you please fix it? </div>       
<br>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    Why isnt it correct? if only by changing the loan purpose and filling with median it made another duplicates it means that these rows are identical in all the other columns!.. it means these customers have the same amount of: children, days employed, dob years, and same: education, family status, gender, income type and debt score... its not logical that there will be different customers with all of those columns be the same... And if you still think i should not drop them then what to do? just leave them?

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing V2:</b><br>
   Usually, duplicate processing is done on raw data, but we filled in the missing values in 2 columns with the median, processed the number of children, and reduced the loan purpose to a consistent look. Thus duplicates were formed that may not actually duplicate, even though the other columns may be identical. It's important to be very careful with deleting data. Especially since you wrote that in real data it could be different clients. And we in our projects are trying to get closer to the real conditions. Better to leave it as it is.</div>

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
   oh, ok then, thats what i was inclined to do from the get go but Anastasia convinced me to remove it for the sake of the drill.. Just like in the first duplicates removal, when i discovered that if i assess and decide what is a duplicate based on 10 columns, then about 3000 rows in the no_NaN table would be duplicates just like the 71 in the NaN table.. anyway i hashed the drop above.
    
   also, just to elaborate what i mean: in a real case like that i assume i would have "client ids" and thus there wont be a need to lean on the other columns to differentiate between customers.

<div style="border:solid green 2px; padding: 20px"> <b>Success V3:</b><br>
   Yes, I agree, when we have the customer IDs, we will be able to tell for sure if they are duplicates or not. But again, it's better to be careful with deleting data so that we don't accidentally delete something important :) </div>   

### Conclusion

I removed the so called duplicates and stemmed the purpose column so now we can start categorizing the data.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   Thanks for conclusion!</div>       

### Categorizing Data

Now ill rename columns. the dob_years column to be more clear and days_employed is not logical so ill assume it is hours which then will make sense.

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
   So since it was established above it cant be hours or minutes (or seconds), i will change days_employed to time_employed instead of hours_employed.

In [59]:
# renaming the dob_years and days_employed columns
renamed_columns = {'dob_years': 'years_in_bank', 'days_employed': 'time_employed', 'family_status': 'marital_status', 'family_status_id': 'marital_status_id'}
orgnzd_credit_score.rename(renamed_columns, axis=1, inplace=True)
orgnzd_credit_score

children  time_employed  years_in_bank            education  \
0             1           8437             42    bachelor's degree   
1             1           4024             36  secondary education   
2             0           5623             33  secondary education   
3             3           4124             32  secondary education   
4             0         340266             53  secondary education   
...         ...            ...            ...                  ...   
21449         1           4529             43  secondary education   
21450         0         343937             67  secondary education   
21451         1           2113             38  secondary education   
21452         3           3112             38  secondary education   
21453         2           1984             40  secondary education   

       education_id     marital_status  marital_status_id gender income_type  \
0                 0            married                  0      F    employee   
1                 1            married                  0      F    employee   
2                 1            married                  0      M    employee   
3                 1            married                  0      M    employee   
4                 1  civil partnership                  1      F     retiree   
...             ...                ...                ...    ...         ...   
21449             1  civil partnership                  1      F    business   
21450             1            married                  0      F     retiree   
21451             1  civil partnership                  1      M    employee   
21452             1            married                  0      M    employee   
21453             1            married                  0      F    employee   

       debt  total_income    purpose  
0         0         40620      house  
1         0         17932        car  
2         0         23341      house  
3         0         42820  education  
4         0         25378    wedding  
...     ...           ...        ...  
21449     0         35966      house  
21450     0         24959        car  
21451     1         14347      house  
21452     1         39054        car  
21453     0         13127        car  

[21454 rows x 12 columns]

### Conclusion

I renamed the columns.

for order's sake ill write all the code need for answering the next questions, below each question respectively.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Good work! 👍</div>    

## Answer these questions

###### childrent/debt relation

- Is there a relation between having kids and repaying a loan on time?

In [60]:
# making a children/debt table
children_debt_relation = orgnzd_credit_score[['children', 'debt']]
children_debt_relation

children  debt
0             1     0
1             1     0
2             0     0
3             3     0
4             0     0
...         ...   ...
21449         1     0
21450         0     0
21451         1     1
21452         3     1
21453         2     0

[21454 rows x 2 columns]

In [61]:
# looking at the value counts for each number of children
children_debt_relation.groupby('children').agg({'debt': ['value_counts']})

debt
              value_counts
children debt             
0        0           13028
         1            1063
1        0            4410
         1             445
2        0            1926
         1             202
3        0             303
         1              27
4        0              37
         1               4
5        0               9

Since there are much less amount of customers that has 1,2,3,4,5 children then customers that dont have at all, there is no point calculating the default rate for each of them separatly because its not an equal comparison, so ill just combine them all in to 2 groups: 'no_kids' and 'have_kids' otherwise we'll get an answer that customers that have 5 kids will never default a loan, but it'll be based only on 9 cases.

In [62]:
# making a function to divide children column in to 2 groups
def kids_group(amount):
    if amount == 0:
        return 'no_kids'
    return 'have_kids'

children_debt_relation['kids'] = children_debt_relation['children'].apply(kids_group)
children_debt_relation

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


children  debt       kids
0             1     0  have_kids
1             1     0  have_kids
2             0     0    no_kids
3             3     0  have_kids
4             0     0    no_kids
...         ...   ...        ...
21449         1     0  have_kids
21450         0     0    no_kids
21451         1     1  have_kids
21452         3     1  have_kids
21453         2     0  have_kids

[21454 rows x 3 columns]

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    All good here.</div>       

In [63]:
# making a pivot table to see how much there is for each group
children_debt_pivot = children_debt_relation.pivot_table(
    index = 'kids',
    #columns = 'debt',
    values = 'debt',
    aggfunc = ['sum', 'count', 'mean']
)
children_debt_pivot

sum  count      mean
           debt   debt      debt
kids                            
have_kids   678   7363  0.092082
no_kids    1063  14091  0.075438

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
   Wow, it's really great that you know this method of pivot table that helps to avoid additional calculations. Very well!</div>       

In [64]:
# adding the default_rate for each group in percents
perc_calc = (children_debt_pivot['mean']) * 100
perc_calc = round(perc_calc, 3)
children_debt_pivot['default_rate'] = (perc_calc).astype(str) + '%'
children_debt_pivot

sum  count      mean default_rate
           debt   debt      debt             
kids                                         
have_kids   678   7363  0.092082       9.208%
no_kids    1063  14091  0.075438       7.544%

### Conclusion

So it seems it would be safer to give a loan to customers with no kids, makes sense, but we also must take into account that the amount of customers with no kids, is almost double the amount of customers with kids so the result is probably somewhat skewed.

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Great!</div>       

###### marital status/debt relation

- Is there a relation between marital status and repaying a loan on time?

In [65]:
# making a marital_status/debt table
marital_status_debt_relation = orgnzd_credit_score[['marital_status', 'debt']]
marital_status_debt_relation

marital_status  debt
0                married     0
1                married     0
2                married     0
3                married     0
4      civil partnership     0
...                  ...   ...
21449  civil partnership     0
21450            married     0
21451  civil partnership     1
21452            married     1
21453            married     0

[21454 rows x 2 columns]

In [66]:
# marital_status_debt_relation.groupby('marital_status').agg({'debt': ['value_counts']})
# making a pivot table to see how much there is for each group
marital_status_debt_pivot = marital_status_debt_relation.pivot_table(
    index = 'marital_status',
    #columns = 'debt',
    values = 'debt',
    aggfunc = ['sum', 'count', 'mean']
)
marital_status_debt_pivot

sum  count      mean
                  debt   debt      debt
marital_status                         
civil partnership  388   4151  0.093471
divorced            85   1195  0.071130
married            931  12339  0.075452
unmarried          274   2810  0.097509
widow / widower     63    959  0.065693

In [67]:
# adding the default_rate for each group in percents
perc_calc = (marital_status_debt_pivot['mean']) * 100
perc_calc = round(perc_calc, 3)
marital_status_debt_pivot['default_rate'] = (perc_calc).astype(str) + '%'
marital_status_debt_pivot

sum  count      mean default_rate
                  debt   debt      debt             
marital_status                                      
civil partnership  388   4151  0.093471       9.347%
divorced            85   1195  0.071130       7.113%
married            931  12339  0.075452       7.545%
unmarried          274   2810  0.097509       9.751%
widow / widower     63    959  0.065693       6.569%

In [68]:
# sorting the table by default_rate
marital_status_debt_pivot = marital_status_debt_pivot.sort_values('default_rate')
marital_status_debt_pivot

sum  count      mean default_rate
                  debt   debt      debt             
marital_status                                      
widow / widower     63    959  0.065693       6.569%
divorced            85   1195  0.071130       7.113%
married            931  12339  0.075452       7.545%
civil partnership  388   4151  0.093471       9.347%
unmarried          274   2810  0.097509       9.751%

### Conclusion

So it seems it would be safest to give a loan to widows, and the riskiest to give to unmarried. i guess it makes sense since a widow is someone who still already has the experience with financial duties and budget (compared to unmarried) but still has the right to their partners assets and money (compared to a divorced person who does not).

(here also we must take into account that the amount of each group is not the same so maybe the result is somewhat skewed)

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    You've done a great job of applying pivot tables to datasets 👍.</div>   

###### income/debt relation

(I answered this question before all the others because i foresaw it will be the most challenging one)

- Is there a relation between income level and repaying a loan on time?

In [69]:
# making a total_income/debt table
income_debt_relation = orgnzd_credit_score[['total_income', 'debt']]
income_debt_relation

total_income  debt
0             40620     0
1             17932     0
2             23341     0
3             42820     0
4             25378     0
...             ...   ...
21449         35966     0
21450         24959     0
21451         14347     1
21452         39054     1
21453         13127     0

[21454 rows x 2 columns]

In [70]:
#income_debt_relation = income_debt_relation.sort_values('total_income').reset_index(drop=True)
#income_debt_relation

In [71]:
# looking at what are the min and max of total_income
income_debt_relation.describe()

total_income          debt
count   21454.000000  21454.000000
mean    26448.305537      0.081150
std     15705.732637      0.273072
min      3306.000000      0.000000
25%     17219.500000      0.000000
50%     22815.000000      0.000000
75%     31330.000000      0.000000
max    362496.000000      1.000000

Since there are a lot of unique incomes i will divide them into 4 groups: low income, below average, above average and high income. I will do it using a function and the apply method.

(I want to stress that since i dont know what currency it is, i have no clue to what is logical to be low income and is the highest absurdly high or not.. because maybe the low incomes comes from someone who has a 25% part time job in cleaning toilets, and the highest from a full time CEO position so the point being i have no indication that way and thats why i am dividing the income to equivalent groups..) - that was my previous plan, but i gave it up because i understood that it isnt logical that exactly the same number of people would have a high income as a low income in a given dataset.

###### So as you can see after a lot of delibaration (between me myself and Anastasia) i decided to make the grouping based on the assumptions that there will be statistically most below_avarage incomes followed by low_income and then above_avarage and lastly the least number of people will have high_income.

In [72]:
## importing numpy
#import numpy as np

#income = np.array(income_debt_relation['total_income'])

#min_income = income_debt_relation['total_income'].min()
#max_income = income_debt_relation['total_income'].max()
#income_50 = (min_income + max_income) / 2
#income_25 = (min_income + income_50) / 2
#income_75 = (max_income + income_50) / 2

# making a function to divide total_income column in to 4 groups
def income_group(amount):
    if amount <= 20000:  # np.percentile(income, 25):    #income_25:     #10000:
        return 'low_income'
    if amount <= 40000:  # np.percentile(income, 50):    #income_50:     #30000:
        return 'below_average'
    if amount <= 60000:  # np.percentile(income, 75):    #income_75:     #90000:
        return 'above_average'
    return 'high_income'

# applying the function to the table and adding the results to a new column
income_debt_relation['income_group'] = income_debt_relation['total_income'].apply(income_group)
income_debt_relation

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


total_income  debt   income_group
0             40620     0  above_average
1             17932     0     low_income
2             23341     0  below_average
3             42820     0  above_average
4             25378     0  below_average
...             ...   ...            ...
21449         35966     0  below_average
21450         24959     0  below_average
21451         14347     1     low_income
21452         39054     1  below_average
21453         13127     0     low_income

[21454 rows x 3 columns]

(The code works even though there is this red notice, i read a lot about what is in the link but there is nothing relevant i found there, and trying its suggestion to "Try using .loc[row_indexer,col_indexer] = value instead" doesnt work and is not relevant.. i am not even using .loc here..)

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    You did well to use the quantile method to segment people by income. The quartile boundaries help to divide the data into equal parts. And equal parts are more convenient to compare with each other :) That's really great! 

Above I wrote to you how to switch off warnings  :)</div>  

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    I did not use percentile in the end because it would not be logical to assume that there are equal amount of customers that earn high income as low income :)

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
    I see. Your idea is logical and has a right to exist, just often use the method of quantiles, which helps divide clients into equal parts to make it easier to compare them with each other. Because if we get a very small group of people, in the end, we can't say anything about it because there won't be enough data to evaluate it.</div>  

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    Yeah i understood it as well thats why i played with the amounts until i got enough people in each group and in logicall proportions

<div style="border:solid green 2px; padding: 20px"> <b>Success V3:</b><br>
    Perfect!</div>  

In [73]:
# checking the count for each group and that it seems logical
income_debt_relation['income_group'].value_counts()

below_average    10886
low_income        7756
above_average     2140
high_income        672
Name: income_group, dtype: int64

Trying to find the default rate

<s><div style="border:solid orange 2px; padding: 20px"> <b>Remarks:</b><br>
    If you are not going to use this code in the future, it is better to delete it.</div>  

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    I thought you would like to see my train of thought..

<s><div style="border:solid orange 2px; padding: 20px"> <b>Remarks V2:</b><br>
   It's great that you use different methods and find the best one for you, but usually in projects it is not accepted to leave code that will not be used. Your work is already good 😊 </div>   

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    removed it

In [74]:
# making a pivot table to see how much there is for each group
income_debt_pivot = income_debt_relation.pivot_table(
    index = ['income_group'],
    #columns = 'debt',
    values = 'debt',
    aggfunc = ['sum', 'count', 'mean']
)

In [75]:
# adding the default_rate for each group in percents
#income_debt_pivot['ratio'] = income_debt_pivot[0] / income_debt_pivot[1]
#income_debt_pivot['0_percentage'] = (income_debt_pivot[0] / (income_debt_pivot[1] + income_debt_pivot[0])) * 100
#perc_calc = (income_debt_pivot[1] / (income_debt_pivot[1] + income_debt_pivot[0])) * 100
perc_calc = (income_debt_pivot['mean']) * 100
perc_calc = round(perc_calc, 3)
income_debt_pivot['default_rate'] = (perc_calc).astype(str) + '%'  #'{perc_calc:.2f}'.format(perc_calc)
income_debt_pivot

sum  count      mean default_rate
              debt   debt      debt             
income_group                                    
above_average  156   2140  0.072897        7.29%
below_average  904  10886  0.083042       8.304%
high_income     38    672  0.056548       5.655%
low_income     643   7756  0.082904        8.29%

In [76]:
# sorting the table by default_rate
income_debt_pivot = income_debt_pivot.sort_values('default_rate')
income_debt_pivot

sum  count      mean default_rate
              debt   debt      debt             
income_group                                    
high_income     38    672  0.056548       5.655%
above_average  156   2140  0.072897        7.29%
low_income     643   7756  0.082904        8.29%
below_average  904  10886  0.083042       8.304%

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Good job! 👏👏👏</div>   

### Conclusion

So it seems it would be safest to give a loan to high_income customers, and the riskiest to give to below_average customers. i guess it makes sense since a customer with a below_average income is someone who's not poor on the one hand so he is used to spending, and on the other hand he still doesnt earn a lot and probably not budget savvy (compared to low_income who despite/because having very little in money is very strict with themselves and have the habbit to budget because of the need to.) and of cource it makes sense that from there the highre the income the safer it is to give the loan, because a higher income ussually comes with more specialized skills and that ussually comes also with maturer person (not necessarily of course) and that in turn comes with more budgeting and finnancial skills.

(here also we must take into account that the amount of each group is not the same so maybe the result is somewhat skewed)

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
Thanks a lot for in-depth analysis!
</div>   

###### purpose/debt relation

- How do different loan purposes affect on-time repayment of the loan?

I first assumed that the purpose column represents the loan the customer wants to get now, not the one he already got from the bank. So which scenario makes more sense based on the question above: 

1 is based on **if the purpose is what the customer asks in the `present`**, 

2 is based on **if the purpose is what the customer asked previously in the `past`**.

I came to the conclusion that the 2nd scenario is the relevant one for this question.

In [77]:
# making a purpose/debt table
purpose_debt_relation = orgnzd_credit_score[['debt', 'purpose']]

In [78]:
#purpose_debt_relation.groupby('purpose').agg({'debt': ['value_counts']})
# making a pivot table to see how much there is for each group
purpose_debt_pivot = purpose_debt_relation.pivot_table(
    index = 'purpose',
    #columns = 'debt',
    values = 'debt',
    aggfunc = ['sum', 'count', 'mean']
) # .reset_index()

purpose_debt_pivot

sum  count      mean
          debt   debt      debt
purpose                        
car        403   4306  0.093590
education  370   4013  0.092200
house      782  10811  0.072334
wedding    186   2324  0.080034

In [79]:
# adding the default_rate for each group in percents
perc_calc = (purpose_debt_pivot['mean']) * 100
perc_calc = round(perc_calc, 3)
purpose_debt_pivot['default_rate'] = (perc_calc).astype(str) + '%'
purpose_debt_pivot

sum  count      mean default_rate
          debt   debt      debt             
purpose                                     
car        403   4306  0.093590       9.359%
education  370   4013  0.092200        9.22%
house      782  10811  0.072334       7.233%
wedding    186   2324  0.080034       8.003%

In [80]:
# sorting the table by default_rate
purpose_debt_pivot = purpose_debt_pivot.sort_values('default_rate')
purpose_debt_pivot

sum  count      mean default_rate
          debt   debt      debt             
purpose                                     
house      782  10811  0.072334       7.233%
wedding    186   2324  0.080034       8.003%
education  370   4013  0.092200        9.22%
car        403   4306  0.093590       9.359%

### Conclusion

So it seems it would be safest to give a loan to house seeking customers, and the riskiest to give to car seeking customers. i guess it makes sense since a car seeking customer will usually need a lot of maintenence for the car and the car will consistently lose value, (compared to education seeking customer who despite having trouble with income while he gets the education, will have a better chance of returning the loan since the education can maybe increase his value [not like a car that only decreases].) and of cource it makes sense that it is safer to give the loan to a house seeking customer, because a house usually increases its value over time and if loses then not much also its a much more serious purchace and therefor usually the customer that asks it is maturer and has more chances of returning the loan on time.

(here also we must take into account that the amount of each group is not the same so maybe the result is somewhat skewed)

<div style="border:solid green 2px; padding: 20px"> <b>Success:</b><br>
    Very well, you went deep into all the questions and wrote logical conclusions based on the results 👍 👍 👍 </div>   

In [81]:
# i copied this "class" way of displaying from the web without understanding it (i tried) because i think its very useful here
# imorting numpy library
import numpy as np

# i just copied the code from here
class display(object):
    """Display HTML representation of multiple objects"""
    template = """<div style="float: left; padding: 10px;">
    <p style='font-family:"Courier New", Courier, monospace'>{0}</p>{1}
    </div>"""
    def __init__(self, *args):
        self.args = args
        
    def _repr_html_(self):
        return '\n'.join(self.template.format(a, eval(a)._repr_html_())
                         for a in self.args)
    
    def __repr__(self):
        return '\n\n'.join(a + '\n' + repr(eval(a))
                           for a in self.args)
# to here
    
display('children_debt_pivot', 'marital_status_debt_pivot', 'income_debt_pivot', 'purpose_debt_pivot')

children_debt_pivot
            sum  count      mean default_rate
           debt   debt      debt             
kids                                         
have_kids   678   7363  0.092082       9.208%
no_kids    1063  14091  0.075438       7.544%

marital_status_debt_pivot
                   sum  count      mean default_rate
                  debt   debt      debt             
marital_status                                      
widow / widower     63    959  0.065693       6.569%
divorced            85   1195  0.071130       7.113%
married            931  12339  0.075452       7.545%
civil partnership  388   4151  0.093471       9.347%
unmarried          274   2810  0.097509       9.751%

income_debt_pivot
               sum  count      mean default_rate
              debt   debt      debt             
income_group                                    
high_income     38    672  0.056548       5.655%
above_average  156   2140  0.072897        7.29%
low_income     643   7756  0.082904        8.29%
below_average  904  10886  0.083042       8.304%

purpose_debt_pivot
           sum  count      mean default_rate
          debt   debt      debt             
purpose                                     
house      782  10811  0.072334       7.233%
wedding    186   2324  0.080034       8.003%
education  370   4013  0.092200        9.22%
car        403   4306  0.093590       9.359%

## General conclusion

###### So my conclusion is: 

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
1 Preprocessing - originally i had data of customers with all sorts of parameters like if they are married, how many children they have what type of education and income amount etc'. the data had problems in form of missing values, errors and duplicates. first i organized the table by lower casing all the relevant object columns, fixed all the negative values and replaced errors like 20 children. then i filled the missing values in total income with the medians by income type (did the same with the NaNs of days employed). then i stemmed the purpose column to have only several big groups, and found and removed duplicate rows. and in the end i got a table of 21120 rows with no NaN and no duplicates or illogical values.

2 Categorization - i decided to split the data into different groups of income to answer the question about the relation of income level to defaulting on a loan. the distribution of income i had is low income (up to 20k), below average(up to 40k), above average(up to 60k) and hgh income(above 60k)

3 Answering questions - 
    for the 1st question with kids group were risky and had higher default-rate and no kids group were trustworthy
    for the 2nd question unmarried group were risky and had higher default-rate and widows/ers group were trustworthy
    for the 3rd question below average group were risky and had higher default-rate and high income group were trustworthy
    for the 4th question car group were risky and had higher default-rate and house group were trustworthy
    My recommendation to the bank, for example, is lowering loan rate for trustworthy good categories

The **`safest`** customers are: `widows/ers` with `no kids`, with `above_average or higher income`, that asking a `loan for a house`.

The **`riskiest`** customers are: `unmarried` people `with kids`, with `below_average or lower income`, that asking a `loan for a car`.

<s><div style="border:solid red 2px; padding: 20px"> <b>Needs fixing:</b><br>
  The final conclusion should be a brief overview of your work, so it's better to describe everything you learned about the data. Please write more about what you researched and what results you got in the process, as the final conclusion is the first thing people read and is often the basis for judging the quality of the work as a whole. And it is not formally accepted to write code in the conclusions, because the conclusions are most often read by customers who don't understand anything about code. Please, fix it. </div>

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
   I like that you wrote such in-depth conclusions on the whole study, proved them with numbers and made logical assumptions. Thank you for your work! It's really great! 👏👏👏</div> 

<div class="alert alert-info" style="border:solid blue 2px; padding: 20px">
    But the code here is only to display all the relevant tables so wht i write can be clearly seen.. its not like i explain something in the code here..

<div style="border:solid green 2px; padding: 20px"> <b>Success V2:</b><br>
   I understand. I like the way you formatted the tables, it's really great! 👏  But usually the general conclusion contains only a textual description of the whole research. </div> 

## Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization; --- I didnt get it.
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.